### DISCUSSION TITLE
How to use Tearing correctly

### DISCUSSION THREAD

**Discussion by RickYang1** [InfinyTech3D/SofaUnity#123] [Help - Using SofaUnity] [answered]
I want to use Tearing to achieve the eyeball capsule membrane tearing. I tested the scene of SofaUnity24_06: `Demo_02/TearingInteraction`, but it's difficult to achieve a similar performance like [this](https://github.com/InfinyTech3D/Tearing)， The performance I tested is as follows:

https://github.com/user-attachments/assets/758c3a5d-0cbf-4d73-a899-6f66486b1462
In my testing, a lot of debris is generated and it is difficult to tear out a continuous path. When I keep dragging a triangle, the performance will be worse(at the 17th second). Why is this? What can I do to optimize it？
Thanks in advance!

**Comment by epernod** [comment, marked as answer]
Hi,

yes we know that the Tearing behavior is not stable. We are working on it but lack of time/resources to improve it.
You can try to modify the `fractureMaxLength` (bigger value will create larger incision each time) and the `stressThreshold` (higher value will create incision only when you stretch more) here:
[in the .scn file](https://github.com/InfinyTech3D/SofaUnity/blob/1ade54ef4dc750c1717c34ffd3abd61d1b4f8259/Scenes/SofaScenes/SquareTissue_02_interaction.scn#L57-L58)

**Reply by RickYang1** [reply]
Thanks for your reply again. I will test the methods you provided.
Best regards!

**Reply by RickYang1** [reply]
Hello,@epernod 
    Regarding the issue of debris generation, my tests have shown that when the cloth undergoes severe deformation, even a single tearing operation can produce jagged and irregular cracks：
https://github.com/user-attachments/assets/0049ff1b-3e85-48bc-8b93-57c172a82f0f
    Upon reviewing the Tearing code, I noticed that in the `m_triangleGeo->computeIntersectionsLineTriangle` function—which computes the intersection between a line and a triangle—the current deformed vertex positions are used. I suspect this is the main reason why jagged cracks occur when the mesh experiences large deformations.
    To address this, I’m considering using the original (undeformed) vertex positions of the mesh to calculate the intersections. For any newly generated vertices, I would compute their corresponding positions in the original mesh using their `BaryCoef` on the intersected edges or triangles.Do you think this approach could effectively solve the problem?
Best regards!

**Reply by epernod** [reply]
the problem on your movie might be related to the snapping threshold. This update the incision path to go through existing vertex to avoid creation of small triangles.
Can you show the parameters you are using in the code and in the scene file.
Also could you tell me which version of the plugin you are using. As I'm currently changing some part of it, I'm a bit lost in the versions

**Reply by RickYang1** [reply]
Yes, of course. Here are the settings of my scene file,：
```
<TearingEngine name="TearingEngine" input_position="@Mo.position" stressThreshold="16.0" 
            step="20" nbFractureMax="100" showFracturePath="1" fractureMaxLength="1" computeVertexStressMethod="0"/>
```
And here is the setting of the snapping threshold in my code (actually, I didn’t make any modifications, and the source code has it set to zero):
```
  int snapingValue = 0;
  int snapingBorderValue = 0;
```
The version of my Tearing is 24.06.
